In [2]:
#Importing libraries:
import requests
import pandas as pd

## Figuring Out How to Collect Data (See End of Notebook for Collection of Data that Was Used for Modeling)

In [76]:
#Selecting URL's to gather info from:

#The url with discussion about climate change (probably mostly believers in climate change):
url_1 = 'https://www.reddit.com/r/climatechange.json'

#The url with skeptic discussion about climate change (probably mostly climate change skeptics):
url_2 = 'https://www.reddit.com/r/climateskeptics.json'

In [77]:
#Creating a header so reddit.com sees a header instead of a computer looking for infomation on the site:
#this helps to get a return res.status_code of 200 instead of 429 error
headers = {'User-agent' : 'Julia bot 0.1'}

In [78]:
#Gathering data from the urls (the data comes in as json objects):
res_1 = requests.get(url_1, headers = headers)
res_2 = requests.get(url_2, headers = headers)

In [79]:
#Checking have 200 rather than 429 error (making sure gathering requests without client error or other errors)
print(res_1.status_code)
print(res_2.status_code)

200
200


In [80]:
#Checking data types of res_1 and res_2
print(type(res_1))
print(type(res_2))

<class 'requests.models.Response'>
<class 'requests.models.Response'>


In [81]:
#Converting response json objects to Python dictionaries
data_1 = res_1.json()
data_2 = res_2.json()

In [82]:
#Checking data types of data_1 and data_2 (should be dictionaries)
print(type(data_1))
print(type(data_2))

<class 'dict'>
<class 'dict'>


In [83]:
#Checking keys of data_1:
data_1.keys()

dict_keys(['kind', 'data'])

In [84]:
#Checking keys of data_2:
data_2.keys()

dict_keys(['kind', 'data'])

In [85]:
#Checking what inside 'kind' key in data_1 (no keys inside)
data_1['kind']

'Listing'

In [86]:
#Checking what inside 'kind' key in data_2 (no keys inside)
data_2['kind']

'Listing'

It looks like there is only information stating that the information in the kind key is the statement that the kind of information in data_1 and data_2 is 'Listing.'  We'll look inside the 'data' key 

In [87]:
#Checking what inside 'data' key in data_1
data_1['data'].keys()

dict_keys(['modhash', 'dist', 'children', 'after', 'before'])

In [88]:
#Checking what inside 'data' key in data_2
data_2['data'].keys()

dict_keys(['modhash', 'dist', 'children', 'after', 'before'])

In [89]:
#Looking at the number of posts collected in data_1:
len(data_1['data']['children'])

27

In [90]:
#Looking at the number of posts collected in data_2:
len(data_2['data']['children'])

25

In [122]:
#Looking at 1st item (index 0) of the posts:
#data_1['data']['children'][0]['data'].keys()

In [126]:
#Checking to see what info in the following location:
data_1['data']['children'][0]['data']['title']

'Subreddit rules'

In [127]:
#Looking at 2nd item (index 1) of the posts:
data_1['data']['children'][1]['data']['title']

"I'm afraid climate change is going to kill me! Help!"

In [128]:
#Looking at 3rd item (index 2) of the posts:
data_1['data']['children'][2]['data']['title']

'More bang for the climate buck: Study identifies hotspots for adaptation funding'

In [129]:
#Looking at 4th item (index 3) of the posts:
data_1['data']['children'][3]['data']['title']

'Alpine glaciers: Another decade of loss'

In [130]:
#Looking at 5th item (index 4) of the posts:
data_1['data']['children'][4]['data']['title']

'Sources of Variation in Climate Sensitivity Estimates (PDF)'

In [131]:
#Looking at 6th item (index 5) of the posts:
data_1['data']['children'][5]['data']['title']

'Has eco-anxiety influenced your parenting/parenthood decisions?'

In [93]:
#Looks like the first two posts are about the subreddit rules and sources to read, respectively,
#so the third item (index 2) and on will be collected for this model:
#data_1['data']['children'][2]

In [181]:
#The url with discussion about climate change (assuming mostly believers in climate change):
url_1 = 'https://www.reddit.com/r/climate.json'

In [182]:
#Setting header to look like Julia bot 0.1 instead of a program when requeting data from the API:
headers = {'User-agent' : 'Julia bot 0.1'}

In [184]:
#Requesting data from url_1 using the header created earlier:
res_1 = requests.get(url_1, headers = headers)

In [185]:
#Checking the status of res_1 (if in 200's good, and if in 400's client-side error)
print(res_1.status_code)

200


In [186]:
#Savind data in json format to data_1:
data_1 = res_1.json()

In [197]:
#Checking length - seeing how many posts pulling in one go:
len(data_1['data']['children'])

26

In [194]:
#Inspecting data_1
#data_1

In [187]:
#Importing libraries for use during while loops when will request a lot of data:
#Noe: decided not to use the clear_output library later on
from IPython.display import clear_output
import time

In [271]:
#Collecting posts on the r/climatechange/new page (note: decided to use r/climate subreddit instead of 
#r/climatechange subreddit)

#Initializing counter:
counter = 0
#Creating a header which will prevent request-related error: 
#(instead of showing up as a computer program looking for data, appearing as "Julia bot 0.3")
headers = {'User-agent': 'Julia bot 0.3'}
#Initializing after variable as True intil it will be replaced with after values in posts:
after = True
#Initializing posts_climate_change list and params dictionary:
posts_climate_change = []
params = {}
#Limiting the loop so 100*25 = 2500 or less posts collected: 
while counter < 100:
    #Making sure that the loop breaks once the final post is collected (when after field is empty):
    if after == None:
        break
    #Collecting posts that will be placed in the posts_climate_change list:
    else:
        #Filling the current post's after value in the params dictionary:
        params = {'after': after}
    #Requesting data from the Reddit API at https://www.reddit.com/r/climatechange/new.json
    #Using the params dictionary to ensure collecting posts sequentially 
    #Using headers so instead of showing up as a computer program looking for data, appearing as "Julia bot 0.3":
    res = requests.get('https://www.reddit.com/r/climatechange/new.json', params, headers=headers)
    #Using the data in the json format:
    the_json = res.json()
    #Getting each full post (with all information related to each post):
    posts_climate_change.extend(the_json['data'].get('children'))
    #Getting the after value for this post so it can be used in params to collect the next post
    after = the_json['data']['after']
    #Adding to counter each loop of the while loop
    counter += 1
    #Time delay so appear as a human clicking through the subreddit instead of a computer
    #(also trying to respect Reddit - don't want to overload their servers)
    time.sleep(1)
    #Printing the length of posts_climate_change to see running total length of 
    #posts_climate_change while this cell runs
    print(len(posts_climate_change))

25
50
75
100
125
150
175
200
225
250
275
300
325
350
375
400
425
450
475
500
525
550
575
600
625
650
675
700
725
750
775
800
825
850
875
900
925
934


In [273]:
#Collecting posts on the r/climate/new page

#Initializing counter:
counter = 0
#Creating a header which will prevent request-related error: 
#(instead of showing up as a computer program looking for data, appearing as "Julia bot 0.3")
headers = {'User-agent': 'Julia bot 0.3'}
#Initializing after variable as True intil it will be replaced with after values in posts:
after = True
#Initializing posts_climate list and params dictionary:
posts_climate = []
params = {}
#Limiting the loop so 100*25 = 2500 or less posts collected: 
while counter < 100:
    #Making sure that the loop breaks once the final post is collected (when after field is empty):
    if after == None:
        break
    #Collecting posts that will be placed in the posts_climate list:
    else:
        #Filling the current post's after value in the params dictionary:
        params = {'after': after}
    #Requesting data from the Reddit API at https://www.reddit.com/r/climate/new.json
    #Using the params dictionary to ensure collecting posts sequentially 
    #Using headers so instead of showing up as a computer program looking for data, appearing as "Julia bot 0.3":
    res = requests.get('https://www.reddit.com/r/climate/new.json', params, headers=headers)
    #Using the data in the json format:
    the_json = res.json()
    #Getting each full post (with all information related to each post):
    posts_climate.extend(the_json['data'].get('children'))
    #Getting the after value for this post so it can be used in params to collect the next post
    after = the_json['data']['after']
    #Adding to counter each loop of the while loop
    counter += 1
    #Time delay so appear as a human clicking through the subreddit instead of a computer
    #(also trying to respect Reddit - don't want to overload their servers)
    time.sleep(1)
    #Printing the length of posts_climate to see running total length of 
    #posts_climate while this cell runs
    print(len(posts_climate))

25
50
75
100
125
150
175
200
225
250
275
300
325
350
375
400
425
450
475
500
525
550
575
600
625
650
675
700
725
750
775
800
825
850
875
900
925
950
975
976


In [274]:
#Finding how many posts_climate collected
len(posts_climate)

976

Note: I'm wondering if 976 posts is enough and why the maximum of 1000 posts were not collected as expected.

In [275]:
#Inspecting title of first post in posts_climate list:
posts_climate[0]['data']['title']

'Which 2020 Democrats Are Strongest on Climate? Here’s the List.'

In [276]:
#Generating list of titles of posts in posts_climate:
post_titles_list_climate = []
for i in range(len(posts_climate)):
    post_title_climate = posts_climate[i]['data']['title']
    post_titles_list_climate.append(post_title_climate)

In [277]:
#Checking have same number of titles in post_titles_list_climate as number of posts in posts_climate:
len(post_titles_list_climate)

976

In [279]:
#Making dataframe with post titles from r/climate/new page:
df_post_titles_climate = pd.DataFrame(post_titles_list_climate)
#Inspecting the head of the df:
df_post_titles_climate.head()

0
0  Which 2020 Democrats Are Strongest on Climate?...
1     Climate Change and the Death of the Small Farm
2  The GOP’s answer to catastrophic climate chang...
3  Ocasio-Cortez’s righteous — and accurate — ang...
4                          Polar bear climate change

In [280]:
#Inspecting the dimensions of df_post_titles_climate:
df_post_titles_climate.shape

(976, 1)

In [281]:
#Collecting posts on the r/climateskeptics/new page

#Initializing counter:
counter = 0
#Creating a header which will prevent request-related error: 
#(instead of showing up as a computer program looking for data, appearing as "Julia bot 0.4")
headers = {'User-agent': 'Julia bot 0.4'}
#Initializing after variable as True intil it will be replaced with after values in posts:
after = True
#Initializing posts_skeptics list and params dictionary:
posts_skeptics = []
params = {}
#Limiting the loop so 100*25 = 2500 or less posts collected:
while counter < 100:
    #Making sure that the loop breaks once the final post is collected (when after field is empty):
    if after == None:
        break
    #Collecting posts that will be placed in the posts_skeptics list:
    else:
        #Filling the current post's after value in the params dictionary:
        params = {'after': after}
    #Requesting data from the Reddit API at https://www.reddit.com/r/climateskeptics/new.json
    #Using the params dictionary to ensure collecting posts sequentially 
    #Using headers so instead of showing up as a computer program looking for data, appearing as "Julia bot 0.4":
    res = requests.get('https://www.reddit.com/r/climateskeptics/new.json', params, headers=headers)
    #Using the data in the json format:
    the_json = res.json()
    #Getting each full post (with all information related to each post):
    posts_skeptics.extend(the_json['data'].get('children'))
    #Getting the after value for this post so it can be used in params to collect the next post
    after = the_json['data']['after']
    #Adding to counter each loop of the while loop
    counter += 1
    #Time delay so appear as a human clicking through the subreddit instead of a computer
    #(also trying to respect Reddit - don't want to overload their servers)
    time.sleep(1)
    #Printing the length of posts_skeptics to see running total length of 
    #posts_skeptics while this cell runs
    print(len(posts_skeptics))

25
50
75
100
125
150
175
200
225
250
275
300
325
350
375
400
425
450
475
500
525
550
575
600
625
650
675
700
725
750
775
800
825
850
875
900
925
950
975
1000
1000


In [324]:
#Collecting posts on the r/climateskeptics/new page
#Note: using len(posts_skeptics) to limit while loop instead of counter here

#Creating a header which will prevent request-related error: 
#(instead of showing up as a computer program looking for data, appearing as "Julia bot 0.4")
headers = {'User-agent': 'Julia bot 0.4'}
#Initializing after variable as True intil it will be replaced with after values in posts:
after = True
#Initializing posts_skeptics list and params dictionary:
posts_skeptics = []
params = {}
#Limiting the loop so 2500 or less posts collected:
while len(posts_skeptics) < 2500:
    #Making sure that the loop breaks once the final post is collected (when after field is empty):
    if after == None:
        break
    #Collecting posts that will be placed in the posts_skeptics list:
    else:
        #Filling the current post's after value in the params dictionary:
        params = {'after': after}
    #Requesting data from the Reddit API at https://www.reddit.com/r/climateskeptics/new.json
    #Using the params dictionary to ensure collecting posts sequentially 
    #Using headers so instead of showing up as a computer program looking for data, appearing as "Julia bot 0.4":
    res = requests.get('https://www.reddit.com/r/climateskeptics/new.json', params, headers=headers)
    #Using the data in the json format:
    the_json = res.json()
    #Getting each full post (with all information related to each post):
    posts_skeptics.extend(the_json['data'].get('children'))
    #Getting the after value for this post so it can be used in params to collect the next post
    after = the_json['data']['after']
    #Time delay so appear as a human clicking through the subreddit instead of a computer
    #(also trying to respect Reddit - don't want to overload their servers)
    time.sleep(1)
    #Printing the length of posts_skeptics to see running total length of 
    #posts_skeptics while this cell runs
    print(len(posts_skeptics))

25
50
75
100
125
150
175
200
225
250
275
300
325
350
375
400
425
450
475
500
525
550
575
600
625
650
675
700
725
750
775
800
825
850
875
900
925
950
975
1000
1000


In [321]:
#Finding number of posts collected from r/climateskeptics/new page:
len(posts_skeptics)

0

In [283]:
#Inspecting the first post's title in posts_skeptics:
posts_skeptics[0]['data']['title']

'IEA: solar’s exponential growth could make it less competitive, not more. New VALCOE metric exposes inherent demand-matching issue'

In [284]:
#Generating list of titles of posts in posts_skeptics:
post_titles_list_skeptics = []
for i in range(len(posts_skeptics)):
    post_title_skeptics = posts_skeptics[i]['data']['title']
    post_titles_list_skeptics.append(post_title_skeptics)

In [285]:
#Checking have same number of titles in post_titles_list_skeptics as posts in posts_skeptics
len(post_titles_list_skeptics)

1000

In [286]:
#Making dataframe with post titles from r/climateskeptics/new page:
df_post_titles_skeptics = pd.DataFrame(post_titles_list_skeptics)
#Inspecting the head of the df:
df_post_titles_skeptics.head()

0
0  IEA: solar’s exponential growth could make it ...
1                   Faith vs Belief 2 Climate Change
2  The climate alarmists are keeping poor people ...
3  Greenland Has Gained 510 Billion Tons of Ice O...
4                    New Video: The Great Flood Myth

In [287]:
#Checking dimensions of df_post_titles_skeptics:
df_post_titles_skeptics.shape

(1000, 1)

In [288]:
#Saving dataframe of r/climate/new post titles to a csv in the data file in this repo:
df_post_titles_climate.to_csv('../data/climate_post_titles_20190328_0915.csv')

In [289]:
#Saving dataframe of r/climateskeptics/new post titles to a csv in the data file in this repo:
df_post_titles_skeptics.to_csv('../data/climateskeptics_post_titles_20190328_0915.csv')

In [311]:
#Collecting full posts in posts_climate_list and 
#post titles in post_titles_climate_list

#Instantiating the lists:
posts_climate_list = []
post_titles_climate_list = []
#Collecting posts and post titles in posts_climate and placing in lists:
for i in range(len(posts_climate)):
    posts_climate_list.append(posts_climate[i])
    post_titles_climate_list.append(posts_climate[i]['data']['title'])

In [312]:
#Finding number of posts in posts_climate_list:
len(posts_climate_list)

976

In [313]:
#Finding number of post titles in post_titles_climate_list (should match the number of posts)
len(post_titles_climate_list)

976

In [316]:
#Instantiating dataframe to place posts and post titles into:
df_posts_climate = pd.DataFrame()
#Adding post_titles_climate_list to 'title' column of df:
df_posts_climate['title'] = post_titles_climate_list
#Adding posts_climate_list to 'full_post' column of df:
df_posts_climate['full_post'] = posts_climate_list
#Inspecting the head of the df:
df_posts_climate.head()

title  \
0  Which 2020 Democrats Are Strongest on Climate?...   
1     Climate Change and the Death of the Small Farm   
2  The GOP’s answer to catastrophic climate chang...   
3  Ocasio-Cortez’s righteous — and accurate — ang...   
4                          Polar bear climate change   

                                           full_post  
0  {'kind': 't3', 'data': {'approved_at_utc': Non...  
1  {'kind': 't3', 'data': {'approved_at_utc': Non...  
2  {'kind': 't3', 'data': {'approved_at_utc': Non...  
3  {'kind': 't3', 'data': {'approved_at_utc': Non...  
4  {'kind': 't3', 'data': {'approved_at_utc': Non...

In [317]:
#Inspecting dimensions of the df:
df_posts_climate.shape

(976, 2)

In [318]:
#Saving dataframe of r/climate/new posts and post titles to a csv in the data file in this repo:
df_posts_climate.to_csv('../data/climate_posts_with_titles_20190328_0915.csv')

In [325]:
#Collecting full posts in posts_skeptics_list and 
#post titles in post_titles_skeptics_list

#Instantiating the lists:
posts_skeptics_list = []
post_titles_skeptics_list = []
#Collecting posts and post titles in posts_skeptics and placing in lists:
for i in range(len(posts_skeptics)):
    posts_skeptics_list.append(posts_skeptics[i])
    post_titles_skeptics_list.append(posts_skeptics[i]['data']['title'])

In [326]:
#Finding number of posts in posts_skeptics_list:
len(posts_skeptics_list)

1000

In [327]:
#Finding number of post titles in post_titles_skeptics_list (should match the number of posts)
len(post_titles_skeptics_list)

1000

In [328]:
#Instantiating dataframe to place posts and post titles into:
df_posts_skeptics = pd.DataFrame()
#Adding post_titles_skeptics_list to 'title' column of df:
df_posts_skeptics['title'] = post_titles_skeptics_list
#Adding posts_climate_list to 'full_post' column of df:
df_posts_skeptics['full_post'] = posts_skeptics_list
#Inspecting the head of the df:
df_posts_skeptics.head()

title  \
0  Professor Valentina Zharkova: The Solar Magnet...   
1  IEA: solar’s exponential growth could make it ...   
2                   Faith vs Belief 2 Climate Change   
3  The climate alarmists are keeping poor people ...   
4  Greenland Has Gained 510 Billion Tons of Ice O...   

                                           full_post  
0  {'kind': 't3', 'data': {'approved_at_utc': Non...  
1  {'kind': 't3', 'data': {'approved_at_utc': Non...  
2  {'kind': 't3', 'data': {'approved_at_utc': Non...  
3  {'kind': 't3', 'data': {'approved_at_utc': Non...  
4  {'kind': 't3', 'data': {'approved_at_utc': Non...

In [329]:
#Saving dataframe of r/climateskeptics/new posts and post titles to a csv in the data file in this repo:
df_posts_skeptics.to_csv('../data/skeptics_posts_with_titles_20190328_0915.csv')

## Attaining Post Comments

In [392]:
#Creating a sample url for testing:
'https://www.reddit.com' + posts_climate[0]['data']['permalink']

'https://www.reddit.com/r/climate/comments/b6iojd/which_2020_democrats_are_strongest_on_climate/'

In [497]:
#Requesting .json data from the url:
res = requests.get('https://www.reddit.com' + posts_climate[0]['data']['permalink']+'.json', headers=headers)

In [498]:
#Ensuring the data is in json format and  
the_json = res.json()

In [534]:
#Inspecting comments within the data (looks like there are layers of comments):
the_json[1]['data']['children'][0]['data']['body']
#len(the_json[1]['data']['children'][0]['data'])
#len(the_json[1]['data']['children'])

"We know Bernie is actually the best, so let's stop pretending like there's a lot of choice and unify behind one strong candidate that already had very good chances last election."

In [556]:
#Creating list of first-level comments on the page (collecting nested replies to these comments requires
#some more work (struggling to figure out how to do this efficiently)
list_i = []
for i in range(len(the_json[1]['data']['children'])):    
    #print(i)
    list_i.append(the_json[1]['data']['children'][i]['data']['body'])

len(list_i)
#Looking at the list to see which comments were collected from the sample url:
list_i


["We know Bernie is actually the best, so let's stop pretending like there's a lot of choice and unify behind one strong candidate that already had very good chances last election.",
 'My "fantasy football" list is Bernie + Andrew Yang as VP= Win  They both coincide on income and environmental factors.  Andrew Yang is considering situations that are twenty years from now.  These are two people that are not just considering what transpires in their terms but what further effects would be.  Edit: too many effects!\n\n']

In [27]:
#Showing tried to find post comments but had some issues, so commented this out:

# headers = {'User-agent': 'Julia bot 0.4'}
# climate_comments_all = []

# for i in range(len(post_titles_climate_list)):
#     res_comments = requests.get('https://www.reddit.com' + 
#                                 posts_climate[i]['data']['permalink'] +'.json', headers=headers)
#     the_json = res_comments.json()
#     comments_each_post = []
#     for j in range(len(the_json[1]['data']['children'])):        
#         if j == []:
#             comments_each_post.append(['No comments'])
#         else:
#             comments_each_post.append(the_json[1]['data']['children'][j]['data']['body'])
        
#         time.sleep(1)
    
#     climate_comments_all.append(comments_each_post)
        
#     clear_output()
#     print(len(climate_comments_all))


In [584]:
#Collecting posts on the r/climate/new page

#Creating a header which will prevent request-related error: 
#(instead of showing up as a computer program looking for data, appearing as "Julia bot 0.3")
headers = {'User-agent': 'Julia bot 0.3'}
#Initializing after variable as True intil it will be replaced with after values in posts:
after = True
#Initializing posts_climate list and params dictionary:
posts_climate = []
params = {}
#Limiting the loop so 2500 or less posts collected: 
while len(posts_climate) < 2500:
    #Making sure that the loop breaks once the final post is collected (when after field is empty):
    if after == None:
        break
    #Collecting posts that will be placed in the posts_climate list:    
    else:
        #Filling the current post's after value in the params dictionary:
        params = {'after': after}
    #Requesting data from the Reddit API at https://www.reddit.com/r/climate/new.json
    #Using the params dictionary to ensure collecting posts sequentially 
    #Using headers so instead of showing up as a computer program looking for data, appearing as "Julia bot 0.3":
    res = requests.get('https://www.reddit.com/r/climate/new.json', params, headers=headers)
    #Using the data in the json format:
    the_json = res.json()
    #Getting each full post (with all information related to each post):
    posts_climate.extend(the_json['data'].get('children'))
    #Getting the after value for this post so it can be used in params to collect the next post
    after = the_json['data']['after']
    #Time delay so appear as a human clicking through the subreddit instead of a computer
    #(also trying to respect Reddit - don't want to overload their servers)
    time.sleep(1)
    #Printing the length of posts_climate to see running total length of 
    #posts_climate while this cell runs
    print(len(posts_climate))

25
50
75
100
125
150
175
200
225
250
275
300
325
350
375
400
425
450
475
500
525
550
575
600
625
650
675
700
725
750
775
800
825
850
875
900
925
950
975
976


In [585]:
#Collecting posts on the r/climateskeptics/new page

#Creating a header which will prevent request-related error: 
#(instead of showing up as a computer program looking for data, appearing as "Julia bot 0.4")
headers = {'User-agent': 'Julia bot 0.4'}
#Initializing after variable as True intil it will be replaced with after values in posts:
after = True
#Initializing posts_skeptics list and params dictionary:
posts_skeptics = []
params = {}
#Limiting the loop so 2500 or less posts collected:
while len(posts_skeptics) < 2500:
    #Making sure that the loop breaks once the final post is collected (when after field is empty):
    if after == None:
        break
    #Collecting posts that will be placed in the posts_skeptics list:
    else:
        #Filling the current post's after value in the params dictionary:
        params = {'after': after}
    #Requesting data from the Reddit API at https://www.reddit.com/r/climateskeptics/new.json
    #Using the params dictionary to ensure collecting posts sequentially 
    #Using headers so instead of showing up as a computer program looking for data, appearing as "Julia bot 0.4":
    res = requests.get('https://www.reddit.com/r/climateskeptics/new.json', params, headers=headers)
    #Using the data in the json format:
    the_json = res.json()
    #Getting each full post (with all information related to each post):
    posts_skeptics.extend(the_json['data'].get('children'))
    #Getting the after value for this post so it can be used in params to collect the next post
    after = the_json['data']['after']
    #Time delay so appear as a human clicking through the subreddit instead of a computer
    #(also trying to respect Reddit - don't want to overload their servers)
    time.sleep(1)
    #Printing the length of posts_skeptics to see running total length of 
    #posts_skeptics while this cell runs
    print(len(posts_skeptics))

25
50
75
100
125
150
175
200
225
250
275
300
325
350
375
400
425
450
475
500
525
550
575
600
625
650
675
700
725
750
775
800
825
850
875
900
925
950
975
999


In [594]:
#Collecting full posts in posts_climate_list and 
#post titles in post_titles_climate_list

posts_climate_list = []
post_titles_climate_list = []
#Collecting posts and post titles in posts_climate and placing in lists:
for i in range(len(posts_climate)):
    posts_climate_list.append(posts_climate[i])
    post_titles_climate_list.append(posts_climate[i]['data']['title'])

In [595]:
#Instantiating dataframe to place posts and post titles into:
df_posts_climate = pd.DataFrame()
#Adding post_titles_climate_list to 'title' column of df:
df_posts_climate['title'] = post_titles_climate_list
#Adding posts_climate_list to 'full_post' column of df:
df_posts_climate['full_post'] = posts_climate_list
#Inspecting the head of the df:
df_posts_climate.head()

title  \
0  'Underwater' homeowners group promotes climate...   
1                Science is not a partisan issue....   
2             Melting glaciers reveal Everest bodies   
3  Climate change could push tropical diseases to...   
4  Which Will Melt First—GOP's Climate Denial, or...   

                                           full_post  
0  {'kind': 't3', 'data': {'approved_at_utc': Non...  
1  {'kind': 't3', 'data': {'approved_at_utc': Non...  
2  {'kind': 't3', 'data': {'approved_at_utc': Non...  
3  {'kind': 't3', 'data': {'approved_at_utc': Non...  
4  {'kind': 't3', 'data': {'approved_at_utc': Non...

In [596]:
#Inspecting dimensions of the df:
df_posts_climate.shape

(976, 2)

In [600]:
#Collecting full posts in posts_skeptics_list and 
#post titles in post_titles_skeptics_list

posts_skeptics_list = []
post_titles_skeptics_list = []
#Collecting posts and post titles in posts_skeptics and placing in lists:
for i in range(len(posts_skeptics)):
    posts_skeptics_list.append(posts_skeptics[i])
    post_titles_skeptics_list.append(posts_skeptics[i]['data']['title'])

In [601]:
#Instantiating dataframe to place posts and post titles into:
df_posts_skeptics = pd.DataFrame()
#Adding post_titles_skeptics_list to 'title' column of df:
df_posts_skeptics['title'] = post_titles_skeptics_list
#Adding posts_skeptics_list to 'full_post' column of df:
df_posts_skeptics['full_post'] = posts_skeptics_list
#Inspecting the head of the df:
df_posts_skeptics.head()

title  \
0  In which cases is the pro-AGW alarmist narrati...   
1  Senator for Utah - Mike Lee's hilarious commen...   
2  Climate science needs a critical review by unb...   
3  Global Warming Has Been Canceled: Key Greenlan...   
4                                A Deal of New Green   

                                           full_post  
0  {'kind': 't3', 'data': {'approved_at_utc': Non...  
1  {'kind': 't3', 'data': {'approved_at_utc': Non...  
2  {'kind': 't3', 'data': {'approved_at_utc': Non...  
3  {'kind': 't3', 'data': {'approved_at_utc': Non...  
4  {'kind': 't3', 'data': {'approved_at_utc': Non...

In [602]:
#Inspecting dimensions of the df:
df_posts_skeptics.shape

(999, 2)

In [603]:
#Saving dataframe of r/climate/new posts and post titles to a csv in the data file in this repo:
df_posts_climate.to_csv('../data/climate_posts_with_titles_20190329_0545.csv')

In [604]:
#Saving dataframe of r/climateskeptics/new posts and post titles to a csv in the data file in this repo:
df_posts_skeptics.to_csv('../data/skeptics_posts_with_titles_20190329_0545.csv')

In [ ]:
#Next try to see if can gather comments again-had an error on evening of 3/38/2019 and unable
#to collect comments so far

In [28]:
#Showing tried to collect r/climate posts and comments, but ran into issues, so commented this out:

# headers = {'User-agent': 'Julia bot 0.4'}
# climate_comments_all = []

# for i in range(len(post_titles_climate_list)):
#     res_comments = requests.get('https://www.reddit.com' + 
#                                 posts_climate[i]['data']['permalink'] +'.json', headers=headers)
#     the_json = res_comments.json()
#     comments_each_post = []
#     for j in range(len(the_json[1]['data']['children'])):        
#         try: 
#             comments_each_post.append(the_json[1]['data']['children'][j]['data']['body'])
#         except j == []:
#             comments_each_post.append(['No comments'])            
        
#         time.sleep(1)
    
#     climate_comments_all.append(comments_each_post)
        
#     clear_output()
#     print(len(climate_comments_all))

## Attaining Post Comment json Data to Unpack Later On 
### (If I decide to use the comments, I will have the data, and I can attain individual comment text and other information from this data)

In [777]:
#Creating header and res for a test run (commented a lot of it because it may be case sensitive - 
#it could cause errors if rerun on a different day when there are no comments on a post, for example)
#headers = {'User-agent': 'Julia bot 0.3'}
#res = requests.get('https://www.reddit.com/r/climate/new.json', headers=headers)

In [778]:
#Making sure no client error when requesting data:
#print(res.status_code)

200


In [779]:
#Using data with json format:
#the_json = res.json()

In [781]:
#Inspecting data keys:
#the_json['data'].keys()

dict_keys(['modhash', 'dist', 'children', 'after', 'before'])

In [780]:
#Inspecting a title:
#the_json['data']['children'][7]['data']['title']

'Great Lakes are rapidly warming, likely to trigger more flooding and extreme weather'

In [757]:
#Inspecting a permalink:
#the_json['data']['children'][7]['data']['permalink']

'/r/climate/comments/b7qdyz/i_was_wrong_on_climate_change_why_cant_other/'

In [758]:
#Inspecting selftext:
#the_json['data']['children'][7]['data']['selftext']

''

In [798]:
#Getting comment data:
# res_comments = requests.get('https://www.reddit.com'+
#                               the_json['data']['children'][7]['data']['permalink'] + '.json',
#                               headers=headers)

In [799]:
#Making sure no client error when requesting data: 
#print(res_comments.status_code)

200


In [800]:
#Using data with json format:
#the_json_1 = res_comments.json()

In [802]:
#Inspecting insides of comments data:

#the_json_1[1]['data'].get('children')
#1st level comments:
#the_json_1[1]['data']['children'][0]['data']['body']
#Next one:
#the_json_1[1]['data']['children'][1]['data']['body']
#len(the_json_1[1]['data']['children'])
#for j in range(len(the_json_1[1]['data']['children'])):
#    print(the_json_1[1]['data']['children'][j]['data']['body'])

In [747]:
#2nd level comments
#the_json_1[1]['data']['children'][0]['data']['replies']['data']['children'][0]['data']['body']
#Next one:
#the_json_1[1]['data']['children'][0]['data']['replies']['data']['children'][0]['data']['body']

'Climate DENIERS are likely to be RACIST. Why? https://www.sierraclub.org/sierra/climate-deniers-are-more-likely-be-racist-obama-trump-climate-change'

In [717]:
#3rd level comments
#the_json_1[1]['data']['children'][1]['data']['replies']['data']['children'][0]['data']['replies']['data']['children'][0]['data']['body']

'But things are changing. If you look back at the amount of past events occurring such as droughts and heatwaves the frequency is increasing, and though the amount of hurricanes is decreasing their intensity is increasing. Flooding was the most common type of hazard last year and caused huge amounts of death and damages in India, Japan, Bangladesh, these being linked to global sea level rise. So things ARE changing, even if it will take a while before everyone is affected by it, so we should do something.'

## Collection of Data that Was Used for Modeling in Next Notebook

#### Collecting Data on 01 Apr 2019

In [822]:
#Collecting subreddit r/climate/new posts

#Creating a header which will prevent request-related error: 
#(instead of showing up as a computer program looking for data, appearing as "Julia bot 0.3")
headers = {'User-agent': 'Julia bot 0.3'}
#Initializing after variable as True intil it will be replaced with after values in posts:
after = True
#Initializing posts_climate list and params dictionary:
posts_climate = []
params = {}
#Limiting the loop so 2500 or less climate posts collected in posts_climate:
while len(posts_climate) < 2500:
    #Making sure that the loop breaks once the final post is collected (when after field is empty):
    if after == None:
        break
    #Collecting posts that will be placed in the posts_climate list:
    else:
        #Filling the current post's after value in the params dictionary:
        params = {'after': after}
    #Requesting data from the Reddit API at https://www.reddit.com/r/climate/new.json
    #Using the params dictionary to ensure collecting posts sequentially 
    #Using headers so instead of showing up as a computer program looking for data, appearing as "Julia bot 0.3":
    res = requests.get('https://www.reddit.com/r/climate/new.json', params, headers=headers)
    #Using the data in the json format:
    the_json = res.json()
    #Getting each full post (with all information related to each post):
    posts_climate.extend(the_json['data'].get('children'))
    #Getting the after value for this post so it can be used in params to collect the next post
    after = the_json['data']['after']
    #Time delay so appear as a human clicking through the subreddit instead of a computer
    #(also trying to respect Reddit - don't want to overload their servers)
    time.sleep(1)
    #Printing the length of posts_climate to see running total length of posts_climate while this cell runs
    print(len(posts_climate))  

25
50
75
100
125
150
175
200
225
250
275
300
325
350
375
400
425
450
475
500
525
550
575
600
625
650
675
700
725
750
775
800
825
850
875
900
925
950
975
977


In [837]:
#Creating lists of post titles, post permalinks, post text (if any text beneath the title), and post comments
#which will be used to build a dataframe of r/climate post information

#Instantiating the lists:
titles_climate = []
permalinks_climate = []
post_text_climate = []
posts_comments_climate = []
#Filling the lists with information from each post in posts_climate:
for i in range(len(posts_climate)):
    #Getting the title of post i:
    titles_climate.append(posts_climate[i]['data']['title'])
    #Getting the permalink of post i:
    permalinks_climate.append(posts_climate[i]['data']['permalink'])
    #Getting the text of post i:
    post_text_climate.append(posts_climate[i]['data']['selftext'])
        
    #Finding the comments for post i:
    #Creating a header which will prevent request-related error: 
    #(instead of showing up as a computer program looking for data, appearing as "Julia bot 0.3")
    headers = {'User-agent': 'Julia bot 0.3'}
    #Getting the after value for this post so it can be used in params to collect the next post
    after = True
    #Initializing list post_i_comments_climate and params dictionary: 
    post_i_comments_climate = []
    params = {}
    #Making sure that the loop breaks once the final post i comment is collected (when after field is empty):
    if after == None:
        break
    #Collecting post i comments that will be placed in the post_i_comments_climate list:
    else:
        #Filling the current comment's after value in the params dictionary:
        params = {'after': after}
    #Requesting data from the Reddit API at https://www.reddit.com/r/climate/new/[post permalink].json
    #Using the params dictionary to ensure collecting post comments sequentially 
    #Using headers so instead of showing up as a computer program looking for data, appearing as "Julia bot 0.3":
    res_comments = requests.get('https://www.reddit.com'+
                                posts_climate[i]['data']['permalink'] + '.json',
                                params,
                                headers=headers)
    #Using the data (comments for post i) in the json format:
    the_json_1 = res_comments.json()
    #Getting each post's full comments page (with all information related to comments for each post):
    post_i_comments_climate.extend(the_json_1[1]['data'].get('children'))
    #Getting the after value for this comment so it can be used in params to collect the next comment:
    after = the_json_1[1]['data']['after']
    #Time delay so appear as a human clicking through the subreddit instead of a computer
    #(also trying to respect Reddit - don't want to overload their servers)
    time.sleep(1)
    #Printing the number of top-level comments (other comments embedded) for each post
    print(len(post_i_comments_climate))
    #Appending post_i_comments_climate list to posts_comments_climate list of comment lists:
    posts_comments_climate.append(post_i_comments_climate)
#Another time delay before collecing next post's comments
time.sleep(1)

0
0
0
0
0
0
1
0
3
3
4
3
0
13
2
2
0
9
2
1
1
0
4
0
1
0
0
0
0
0
8
1
5
5
1
5
2
0
1
0
5
1
0
1
2
0
3
1
1
3
0
2
4
1
2
0
5
2
1
2
0
1
0
2
3
4
2
3
5
6
1
0
0
4
0
2
1
0
2
3
4
0
1
2
0
0
0
0
0
1
0
4
0
6
3
0
7
1
0
1
1
0
0
0
0
1
0
0
0
3
3
0
3
3
0
0
1
1
1
2
1
5
1
0
0
1
0
0
0
5
0
1
5
0
0
3
1
1
3
0
0
3
2
1
1
0
0
1
0
0
0
4
6
0
0
2
1
0
1
1
0
3
1
0
0
1
0
0
0
0
1
0
2
0
0
1
1
3
1
1
0
0
2
1
2
0
2
0
0
4
0
3
3
1
0
0
0
9
1
0
4
2
3
0
0
0
1
4
0
0
1
1
1
0
1
0
3
1
0
0
1
0
1
8
1
7
4
2
1
0
2
3
1
0
1
4
8
5
0
1
1
11
9
2
1
1
1
0
3
1
0
0
0
0
0
1
3
1
2
4
0
3
1
5
2
0
0
0
1
1
0
0
1
1
4
2
2
0
0
3
0
1
1
1
0
0
0
0
2
0
2
1
4
0
0
0
0
0
0
0
0
0
0
4
4
0
0
0
0
0
0
3
1
1
1
0
0
3
0
3
5
1
1
0
0
0
0
1
8
2
0
2
1
3
0
0
0
2
2
1
1
0
0
0
1
0
1
0
4
6
1
0
2
2
1
4
0
0
1
2
3
2
6
1
0
1
0
0
0
1
2
7
0
0
0
1
0
0
2
7
1
0
1
1
0
0
0
2
0
1
0
0
0
4
2
1
6
0
0
1
3
1
4
4
3
6
4
1
6
0
0
1
0
0
5
7
4
0
0
3
0
7
13
0
2
0
0
7
1
2
1
1
0
3
5
0
0
0
1
5
4
2
0
2
0
0
2
0
0
0
0
4
0
4
0
0
0
1
4
2
4
8
0
1
0
5
2
3
0
4
0
1
0
0
2
0
0
1
6
1
1
2
3
1
0
0
3
0
1
0
1
0
0
0
0
13
0
2


In [840]:
#Checking length of titles_climate list (should be same as length of posts_climate)
print(len(titles_climate))

977


In [841]:
#Checking length of posts_climate
print(len(posts_climate))

977


In [842]:
#Checking length of permalinks_climate list (should be same as length of posts_climate)
print(len(permalinks_climate))

977


In [844]:
#Checking length of post_text_climate (should be same as length of posts_climate)
print(len(post_text_climate))

977


In [845]:
#Checking length of posts_comments_climate
print(len(posts_comments_climate))

10747


In [846]:
#Instantiating dataframe to place posts, post titles, post permalinks, and post comments into:
df_posts_climate = pd.DataFrame()
#Adding post_titles_climate_list to 'title' column of df:
df_posts_climate['title'] = titles_climate
#Adding posts_climate to 'full_post' column of df:
df_posts_climate['full_post'] = posts_climate
#Adding permalinks_climate to 'permalink' column of df:
df_posts_climate['permalink'] = permalinks_climate
#Adding post_text_climate to 'post_text' column of df:
df_posts_climate['post_text'] = post_text_climate
#Adding posts_comments_climate to 'full_comments_page_each_post' column of df:
df_posts_climate['full_comments_page_each_post'] = posts_comments_climate
#Inspecting the head of the df:
df_posts_climate.head()

title  \
0  What is anomalous? The stabilisation of CH₄ co...   
1  Global carbon emissions hit record high in 201...   
2  U.S. judge scraps Trump order opening Arctic, ...   
3  Ocasio-Cortez slams critics who would wait on ...   
4                                   Save the climate   

                                           full_post  \
0  {'kind': 't3', 'data': {'approved_at_utc': Non...   
1  {'kind': 't3', 'data': {'approved_at_utc': Non...   
2  {'kind': 't3', 'data': {'approved_at_utc': Non...   
3  {'kind': 't3', 'data': {'approved_at_utc': Non...   
4  {'kind': 't3', 'data': {'approved_at_utc': Non...   

                                           permalink  \
0  /r/climate/comments/b8364j/what_is_anomalous_t...   
1  /r/climate/comments/b82tsw/global_carbon_emiss...   
2  /r/climate/comments/b82cw2/us_judge_scraps_tru...   
3  /r/climate/comments/b826lv/ocasiocortez_slams_...   
4       /r/climate/comments/b8214y/save_the_climate/   

                                           post_text  \
0                                                      
1                                                      
2                                                      
3                                                      
4  Ok, we need to save the climate. Start to post...   

  full_comments_page_each_post  
0                           []  
1                           []  
2                           []  
3                           []  
4                           []

In [847]:
#Checking dimensions of df_posts_climate (should be length of posts_climate, 5):
df_posts_climate.shape

(977, 5)

In [848]:
#Saving dataframe of r/climate/new posts, post titles, post permalinks, and post comments to a csv 
#in the data file in this repo:
df_posts_climate.to_csv('../data/climate_posts_detailed_20190401_0955.csv')

In [820]:
#Making sure will be able to unpack comments later on:

#Getting level 1 comments for each post:
#post_i_comments_level_1 = []
#for j in range(len(the_json_1[1]['data']['children'])):
#    try:
#        post_i_comments_level_1.append(the_json_1[1]['data']['children'][j]['data']['body'])
#    except:
#        post_i_comments_level_1.append("None")
#Adding each post's list of 1st level comments to post_comments_level_1
#post_comments_level_1.extend(post_i_comments_level_1)

0
0
0
0
1
5


In [849]:
#Collecting posts on the r/climateskeptics/new page

#Creating a header which will prevent request-related error: 
#(instead of showing up as a computer program looking for data, appearing as "Julia bot 0.3")
headers = {'User-agent': 'Julia bot 0.3'}
#Initializing after variable as True intil it will be replaced with after values in posts:
after = True
#Initializing posts_skeptics list and params dictionary:
posts_skeptics = []
params = {}
#Limiting the loop so 2500 or less posts collected:
while len(posts_skeptics) < 2500:
    #Making sure that the loop breaks once the final post is collected (when after field is empty):
    if after == None:
        break
    #Collecting posts that will be placed in the posts_skeptics list:
    else:
        #Filling the current post's after value in the params dictionary:
        params = {'after': after}
    #Requesting data from the Reddit API at https://www.reddit.com/r/climateskeptics/new.json
    #Using the params dictionary to ensure collecting posts sequentially 
    #Using headers so instead of showing up as a computer program looking for data, appearing as "Julia bot 0.3":
    res = requests.get('https://www.reddit.com/r/climateskeptics/new.json', params, headers=headers)
    #Using the data in the json format:
    the_json = res.json()
    #Getting each full post (with all information related to each post):
    posts_skeptics.extend(the_json['data'].get('children'))
    #Getting the after value for this post so it can be used in params to collect the next post
    after = the_json['data']['after']
    #Time delay so appear as a human clicking through the subreddit instead of a computer
    #(also trying to respect Reddit - don't want to overload their servers)
    time.sleep(1)
    #Printing the length of posts_skeptics to see running total length of 
    #posts_skeptics while this cell runs
    print(len(posts_skeptics))  

25
50
75
100
125
150
175
200
225
250
275
300
325
350
375
400
425
450
475
500
525
550
575
600
625
650
675
700
725
750
775
800
825
850
875
900
925
950
975
999


In [850]:
#Creating lists of post titles, post permalinks, post text (if any text beneath the title), and post comments
#which will be used to build a dataframe of r/climateskeptics post information

#Instantiating the lists:
titles_skeptics = []
permalinks_skeptics = []
post_text_skeptics = []
posts_comments_skeptics = []
#Filling the lists with information from each post in posts_skeptics:
for i in range(len(posts_skeptics)):
    #Getting the title of post i:
    titles_skeptics.append(posts_skeptics[i]['data']['title'])
    #Getting the permalink of post i:
    permalinks_skeptics.append(posts_skeptics[i]['data']['permalink'])
    #Getting the text of post i:
    post_text_skeptics.append(posts_skeptics[i]['data']['selftext'])
        
    #Finding the comments for post i:
    #Creating a header which will prevent request-related error: 
    #(instead of showing up as a computer program looking for data, appearing as "Julia bot 0.3")
    headers = {'User-agent': 'Julia bot 0.3'}
    #Getting the after value for this post so it can be used in params to collect the next post
    after = True
    #Initializing list post_i_comments_skeptics and params dictionary: 
    post_i_comments_skeptics = []
    params = {}
    #Making sure that the loop breaks once the final post i comment is collected (when after field is empty):
    if after == None:
        break
    #Collecting post i comments that will be placed in the post_i_comments_skeptics list:
    else:
        #Filling the current comment's after value in the params dictionary:
        params = {'after': after}
    #Requesting data from the Reddit API at https://www.reddit.com/r/climateskeptics/new/[post permalink].json
    #Using the params dictionary to ensure collecting post comments sequentially 
    #Using headers so instead of showing up as a computer program looking for data, appearing as "Julia bot 0.3":
    res_comments = requests.get('https://www.reddit.com'+
                                posts_skeptics[i]['data']['permalink'] + '.json',
                                params,
                                headers=headers)
    #Using the data (comments for post i) in the json format:
    the_json_1 = res_comments.json()
    #Getting each post's full comments page (with all information related to comments for each post):
    post_i_comments_skeptics.extend(the_json_1[1]['data'].get('children'))
    #Getting the after value for this comment so it can be used in params to collect the next comment:
    after = the_json_1[1]['data']['after']
    #Time delay so appear as a human clicking through the subreddit instead of a computer
    #(also trying to respect Reddit - don't want to overload their servers)
    time.sleep(1)
    #Printing the number of top-level comments (other comments embedded) for each post
    print(len(post_i_comments_skeptics)) 
    #Appending post_i_comments_skeptics list to posts_comments_skeptics list of comment lists:
    posts_comments_skeptics.append(post_i_comments_skeptics)
#Another time delay before collecing next post's comments
time.sleep(1)

2
4
4
0
0
7
6
0
2
0
4
8
2
5
3
7
1
1
6
0
6
6
2
5
2
0
1
1
1
4
5
0
0
0
1
5
0
3
1
1
6
1
1
1
0
2
2
6
3
2
0
2
2
5
10
0
0
2
5
1
0
3
2
5
1
1
5
7
8
1
0
1
0
5
2
0
1
5
2
0
5
2
3
3
2
1
3
2
1
1
1
1
1
6
1
0
4
4
1
0
0
0
1
1
2
1
1
2
1
3
3
1
0
3
3
1
3
1
4
2
0
1
0
2
0
3
8
1
2
2
2
2
2
1
2
14
1
3
0
10
0
0
17
2
1
1
9
0
4
3
1
2
1
1
4
6
2
6
4
2
7
0
2
1
0
1
14
0
1
0
1
10
9
1
1
2
3
4
2
4
2
0
16
4
1
5
6
0
4
1
10
1
1
4
2
6
2
9
2
1
0
4
1
6
0
7
0
0
0
17
0
0
0
2
1
0
1
4
1
1
0
1
2
2
1
2
7
2
0
9
0
2
1
0
8
0
3
5
1
1
0
3
4
3
3
3
3
1
7
9
5
9
2
2
2
1
4
0
3
5
4
7
0
2
2
6
0
0
0
3
0
1
1
2
3
1
1
4
1
2
1
0
2
5
9
7
3
1
1
7
1
1
0
3
1
3
8
1
0
1
0
3
4
1
1
3
2
7
3
0
7
2
1
6
1
3
3
3
2
1
3
2
8
2
6
3
1
1
2
9
2
6
2
2
1
5
1
3
1
0
10
2
2
2
1
8
1
2
4
1
5
4
4
1
8
0
6
2
5
1
5
2
0
0
5
2
0
1
3
1
6
2
2
3
2
1
2
2
2
0
5
1
4
0
33
5
2
3
2
2
2
3
1
0
3
4
3
2
4
1
2
2
3
3
2
4
1
4
2
3
1
2
0
14
0
18
1
3
1
4
3
0
1
9
4
3
4
2
2
2
4
3
2
9
1
3
3
8
7
1
5
1
0
5
2
3
6
0
7
2
8
1
10
1
4
0
2
2
4
0
1
0
18
6
2
0
11
1
2
2
0
4
6
2
1
1
0
3
0
4
2
0
3
9
1
1
2
1
3
1
1
4


In [851]:
#Checking length of posts_comments_skeptics same as length of posts_skeptics:
len(posts_comments_skeptics)

999

In [855]:
#Instantiating dataframe to place posts, post titles, post permalinks, and post comments into:
df_posts_skeptics = pd.DataFrame()
#Adding titles_skeptics to 'title' column of df:
df_posts_skeptics['title'] = titles_skeptics
#Adding posts_skeptics to 'full_post' column of df:
df_posts_skeptics['full_post'] = posts_skeptics
#Adding permalinks_skeptics to 'permalink' column of df:
df_posts_skeptics['permalink'] = permalinks_skeptics
#Adding post_text_skeptics to 'post_text' column of df:
df_posts_skeptics['post_text'] = post_text_skeptics
#Adding posts_comments_skeptics to 'full_comments_page_each_post' column of df:
df_posts_skeptics['full_comments_page_each_post'] = posts_comments_skeptics
#Inspecting the head of the df:
df_posts_skeptics.head()

title  \
0  Climate Change Alarmists Routinely Ignore Chin...   
1  The Conversation: Lets Create a Youth Climate ...   
2  NASA Now Altering ‘Unadjusted” Data To Create ...   
3  Australia: Labor tries to have it both ways on...   
4       I made a book list. Feel free to contribute.   

                                           full_post  \
0  {'kind': 't3', 'data': {'approved_at_utc': Non...   
1  {'kind': 't3', 'data': {'approved_at_utc': Non...   
2  {'kind': 't3', 'data': {'approved_at_utc': Non...   
3  {'kind': 't3', 'data': {'approved_at_utc': Non...   
4  {'kind': 't3', 'data': {'approved_at_utc': Non...   

                                           permalink post_text  \
0  /r/climateskeptics/comments/b82s1h/climate_cha...             
1  /r/climateskeptics/comments/b7zono/the_convers...             
2  /r/climateskeptics/comments/b7x0qk/nasa_now_al...             
3  /r/climateskeptics/comments/b7w3h7/australia_l...             
4  /r/climateskeptics/comments/b7v3qo/i_made_a_bo...             

                        full_comments_page_each_post  
0  [{'kind': 't1', 'data': {'subreddit_id': 't5_2...  
1  [{'kind': 't1', 'data': {'subreddit_id': 't5_2...  
2  [{'kind': 't1', 'data': {'subreddit_id': 't5_2...  
3                                                 []  
4                                                 []

In [853]:
#Saving dataframe of r/climateskeptics/new posts, post titles, post permalinks, and post comments to a csv 
#in the data file in this repo:
df_posts_skeptics.to_csv('../data/skeptics_posts_detailed_20190401_1040.csv')

#### Collecting Data on 02 Apr 2019

In [5]:
#Importing libraries just in case:

import requests
import time
import pandas as pd

In [7]:
#Collecting subreddit r/climate/new posts

#Creating a header which will prevent request-related error: 
#(instead of showing up as a computer program looking for data, appearing as "Julia bot 0.3")
headers = {'User-agent': 'Julia bot 0.3'}
#Initializing after variable as True intil it will be replaced with after values in posts:
after = True
#Initializing posts_climate list and params dictionary:
posts_climate = []
params = {}

#Limiting the loop so 2500 or less climate posts collected in posts_climate: 
while len(posts_climate) < 2500:
    #Making sure that the loop breaks once the final post is collected (when after field is empty):
    if after == None:
        break
    #Collecting posts that will be placed in the posts_climate list:
    else:
        #Filling the current post's after value in the params dictionary:
        params = {'after': after}
    #Requesting data from the Reddit API at https://www.reddit.com/r/climate/new.json
    #Using the params dictionary to ensure collecting posts sequentially 
    #Using headers so instead of showing up as a computer program looking for data, appearing as "Julia bot 0.3":
    res = requests.get('https://www.reddit.com/r/climate/new.json', params, headers=headers)
    #Using the data in the json format:
    the_json = res.json()
    #Getting each full post (with all information related to each post):
    posts_climate.extend(the_json['data'].get('children'))
    #Getting the after value for this post so it can be used in params to collect the next post
    after = the_json['data']['after']
    #Time delay so appear as a human clicking through the subreddit instead of a computer
    #(also trying to respect Reddit - don't want to overload their servers)
    time.sleep(1)
    #Printing the length of posts_climate to see running total length of posts_climate while this cell runs
    print(len(posts_climate))    

25
50
75
100
125
150
175
200
225
250
275
300
325
350
375
400
425
450
475
500
525
550
575
600
625
650
675
700
725
750
775
800
825
850
875
900
925
950
975
976


In [8]:
#Creating lists of post titles, post permalinks, post text (if any text beneath the title), and post comments
#which will be used to build a dataframe of r/climate post information

#Instantiating the lists:
titles_climate = []
permalinks_climate = []
post_text_climate = []
posts_comments_climate = []
#Filling the lists with information from each post in posts_climate:
for i in range(len(posts_climate)):
    #Getting the title of post i:
    titles_climate.append(posts_climate[i]['data']['title'])
    #Getting the permalink of post i:
    permalinks_climate.append(posts_climate[i]['data']['permalink'])
    #Getting the text of post i:
    post_text_climate.append(posts_climate[i]['data']['selftext'])
        
    #Finding the comments for post i:
    #Creating a header which will prevent request-related error: 
    #(instead of showing up as a computer program looking for data, appearing as "Julia bot 0.3")
    headers = {'User-agent': 'Julia bot 0.3'}
    #Getting the after value for this post so it can be used in params to collect the next post
    after = True
    #Initializing list post_i_comments_climate and params dictionary: 
    post_i_comments_climate = []
    params = {}
    #Making sure that the loop breaks once the final post i comment is collected (when after field is empty):
    if after == None:
        break
    #Collecting post i comments that will be placed in the post_i_comments_climate list:
    else:
        #Filling the current comment's after value in the params dictionary:
        params = {'after': after}
    #Requesting data from the Reddit API at https://www.reddit.com/r/climate/new/[post permalink].json
    #Using the params dictionary to ensure collecting post comments sequentially 
    #Using headers so instead of showing up as a computer program looking for data, appearing as "Julia bot 0.3":
    res_comments = requests.get('https://www.reddit.com'+
                                posts_climate[i]['data']['permalink'] + '.json',
                                params,
                                headers=headers)
    #Using the data (comments for post i) in the json format:
    the_json_1 = res_comments.json()
    #Getting each post's full comments page (with all information related to comments for each post):
    post_i_comments_climate.extend(the_json_1[1]['data'].get('children'))
    #Getting the after value for this comment so it can be used in params to collect the next comment:
    after = the_json_1[1]['data']['after']
    #Time delay so appear as a human clicking through the subreddit instead of a computer
    #(also trying to respect Reddit - don't want to overload their servers)
    time.sleep(1)
    #Printing the number of top-level comments (other comments embedded) for each post:
    print(len(post_i_comments_climate)) 
    #Appending post_i_comments_climate list to posts_comments_climate list of comment lists:
    posts_comments_climate.append(post_i_comments_climate)
#Another time delay before collecing next post's comments
time.sleep(1)

2
0
0
2
1
3
1
0
0
0
1
5
0
0
0
0
1
0
0
2
7
0
0
0
0
3
0
0
0
0
0
3
0
1
1
0
3
3
4
3
0
13
2
2
0
9
2
1
1
0
4
0
1
0
0
0
0
0
8
1
5
5
1
5
2
0
1
0
5
1
0
1
2
0
3
1
1
3
0
2
4
1
2
0
5
3
1
2
0
1
0
2
3
4
2
3
5
6
1
0
0
4
0
2
1
0
2
3
4
0
1
2
0
0
0
0
0
1
0
4
0
6
3
0
7
1
0
1
1
0
0
0
0
1
0
0
0
3
3
0
3
3
0
0
1
1
1
2
1
5
1
0
0
1
0
0
0
5
0
1
5
0
0
3
1
1
3
0
0
3
2
1
1
0
0
1
0
0
0
4
6
0
0
2
1
0
1
1
0
3
1
0
0
1
0
0
0
0
1
0
2
0
0
1
1
3
1
1
0
0
2
1
2
0
2
0
0
4
0
3
3
1
0
0
0
9
1
0
4
2
3
0
0
0
1
4
0
0
1
1
1
0
1
0
3
1
0
0
1
0
1
8
1
7
4
2
1
0
2
3
1
0
1
4
8
5
0
1
1
11
9
2
1
1
1
0
3
1
0
0
0
0
0
1
1
2
4
0
3
1
5
2
0
0
0
1
1
0
0
1
1
4
2
2
0
0
3
0
1
1
1
0
0
0
0
2
0
2
1
4
0
0
0
0
0
0
0
0
0
0
4
4
0
0
0
0
0
0
3
1
1
1
0
0
3
0
3
5
1
1
0
0
0
0
1
8
2
0
2
1
3
0
0
0
2
2
1
1
0
0
0
1
0
1
0
4
6
1
0
2
2
1
4
0
0
1
2
3
2
6
1
0
1
0
0
0
1
2
7
0
0
0
1
0
0
2
7
1
0
1
1
0
0
0
2
0
1
0
0
0
4
2
1
6
0
0
1
3
1
4
4
3
6
4
1
6
0
0
1
0
0
5
7
4
0
0
3
0
7
13
0
2
0
0
7
1
2
1
1
0
3
5
0
0
0
1
5
4
2
0
2
0
0
2
0
0
0
0
4
0
4
0
0
0
1
4
2
4
8
0
1
0
5
2
3
0
4
0
1

In [9]:
#Instantiating dataframe to place posts, post titles, post permalinks, and post comments into:
df_posts_climate = pd.DataFrame()
#Adding titles_climate to 'title' column of df:
df_posts_climate['title'] = titles_climate
#Adding posts_climate to 'full_post' column of df:
df_posts_climate['full_post'] = posts_climate
#Adding permalinks_climate to 'permalink' column of df:
df_posts_climate['permalink'] = permalinks_climate
#Adding post_text_climate to 'post_text' column of df:
df_posts_climate['post_text'] = post_text_climate
#Adding posts_comments_climate to 'full_comments_page_each_post' column of df:
df_posts_climate['full_comments_page_each_post'] = posts_comments_climate
#Inspecting the head of the df:
df_posts_climate.head()

title  \
0  Resources to objectively debate with people wh...   
1  A World Without Water. A short film inspired b...   
2  Polar Warning: Even Antarctica’s Coldest Regio...   
3  Chemical distributor makes tasteless April Foo...   
4  Shell quits major US oil lobby over climate ch...   

                                           full_post  \
0  {'kind': 't3', 'data': {'approved_at_utc': Non...   
1  {'kind': 't3', 'data': {'approved_at_utc': Non...   
2  {'kind': 't3', 'data': {'approved_at_utc': Non...   
3  {'kind': 't3', 'data': {'approved_at_utc': Non...   
4  {'kind': 't3', 'data': {'approved_at_utc': Non...   

                                           permalink  \
0  /r/climate/comments/b8r4rw/resources_to_object...   
1  /r/climate/comments/b8qb80/a_world_without_wat...   
2  /r/climate/comments/b8pl16/polar_warning_even_...   
3  /r/climate/comments/b8o1ug/chemical_distributo...   
4  /r/climate/comments/b8nt6l/shell_quits_major_u...   

                                           post_text  \
0  What resources can I read/view to easily under...   
1                                                      
2                                                      
3                                                      
4                                                      

                        full_comments_page_each_post  
0  [{'kind': 't1', 'data': {'subreddit_id': 't5_2...  
1                                                 []  
2                                                 []  
3  [{'kind': 't1', 'data': {'subreddit_id': 't5_2...  
4  [{'kind': 't1', 'data': {'subreddit_id': 't5_2...

In [10]:
#Saving dataframe of r/climate/new posts, post titles, post permalinks, and post comments to a csv 
#in the data file in this repo:
df_posts_climate.to_csv('../data/climate_posts_detailed_20190402_2100.csv')

In [20]:
#Collecting subreddit r/climateskeptics/new posts

#Creating a header which will prevent request-related error: 
#(instead of showing up as a computer program looking for data, appearing as "Julia bot 0.3")
headers = {'User-agent': 'Julia bot 0.3'}
#Initializing after variable as True intil it will be replaced with after values in posts:
after = True
#Initializing posts_skeptics list and params dictionary:
posts_skeptics = []
params = {}

#Limiting the loop so 2500 or less climateskeptics posts collected in posts_skeptics: 
while len(posts_skeptics) < 2500:
    #Making sure that the loop breaks once the final post is collected (when after field is empty):
    if after == None:
        break
    #Collecting posts that will be placed in the posts_skeptics list:
    else:
        #Filling the current post's after value in the params dictionary:
        params = {'after': after}
    #Requesting data from the Reddit API at https://www.reddit.com/r/climateskeptics/new.json
    #Using the params dictionary to ensure collecting posts sequentially 
    #Using headers so instead of showing up as a computer program looking for data, appearing as "Julia bot 0.3":
    res = requests.get('https://www.reddit.com/r/climateskeptics/new.json', params, headers=headers)
    #Using the data in the json format:
    the_json = res.json()
    #Getting each full post (with all information related to each post):
    posts_skeptics.extend(the_json['data'].get('children'))
    #Getting the after value for this post so it can be used in params to collect the next post
    after = the_json['data']['after']
    #Time delay so appear as a human clicking through the subreddit instead of a computer
    #(also trying to respect Reddit - don't want to overload their servers)
    time.sleep(1)
    #Printing the length of posts_skeptics to see running total length of posts_skeptics while this cell runs
    print(len(posts_skeptics))  

25
50
75
100
125
150
175
200
225
250
275
300
325
350
375
400
425
450
475
500
525
550
575
600
625
650
675
700
725
750
775
800
825
850
875
900
925
950
975
1000
1000


In [21]:
#Creating lists of post titles, post permalinks, post text (if any text beneath the title), and post comments
#which will be used to build a dataframe of r/climateskeptics post information

#Instantiating the lists:
titles_skeptics = []
permalinks_skeptics = []
post_text_skeptics = []
posts_comments_skeptics = []
#Filling the lists with information from each post in posts_skeptics:
for i in range(len(posts_skeptics)):
    #Getting the title of post i:
    titles_skeptics.append(posts_skeptics[i]['data']['title'])
    #Getting the permalink of post i:
    permalinks_skeptics.append(posts_skeptics[i]['data']['permalink'])
    #Getting the text of post i:
    post_text_skeptics.append(posts_skeptics[i]['data']['selftext'])
        
    #Finding the comments for post i:
    #Creating a header which will prevent request-related error: 
    #(instead of showing up as a computer program looking for data, appearing as "Julia bot 0.3")
    headers = {'User-agent': 'Julia bot 0.3'}
    #Getting the after value for this post so it can be used in params to collect the next post
    after = True
    #Initializing list post_i_comments_skeptics and params dictionary: 
    post_i_comments_skeptics = []
    params = {}
    #Making sure that the loop breaks once the final post i comment is collected (when after field is empty):
    if after == None:
        break
    #Collecting post i comments that will be placed in the post_i_comments_skeptics list:
    else:
        #Filling the current comment's after value in the params dictionary:
        params = {'after': after}
    #Requesting data from the Reddit API at https://www.reddit.com/r/climateskeptics/new/[post permalink].json
    #Using the params dictionary to ensure collecting post comments sequentially 
    #Using headers so instead of showing up as a computer program looking for data, appearing as "Julia bot 0.3":
    res_comments = requests.get('https://www.reddit.com'+
                                posts_skeptics[i]['data']['permalink'] + '.json',
                                params,
                                headers=headers)
    #Using the data (comments for post i) in the json format:
    the_json_1 = res_comments.json()
    #Getting each post's full comments page (with all information related to comments for each post):
    post_i_comments_skeptics.extend(the_json_1[1]['data'].get('children'))
    #Getting the after value for this comment so it can be used in params to collect the next comment:
    after = the_json_1[1]['data']['after']
    #Time delay so appear as a human clicking through the subreddit instead of a computer
    #(also trying to respect Reddit - don't want to overload their servers)
    time.sleep(1)
    #Printing the number of top-level comments (other comments embedded) for each post
    print(len(post_i_comments_skeptics)) 
    #Appending post_i_comments_skeptics list to posts_comments_skeptics list of comment lists:
    posts_comments_skeptics.append(post_i_comments_skeptics)
#Another time delay before collecing next post's comments
time.sleep(1)

1
1
1
0
1
1
5
1
5
3
5
1
1
1
3
0
4
4
7
0
0
7
6
0
2
0
4
8
2
5
3
7
1
1
6
0
6
6
2
5
2
0
1
1
1
5
5
0
0
0
1
5
0
3
1
1
6
1
1
1
0
2
2
6
3
2
0
2
2
5
10
0
0
2
5
1
0
3
2
5
1
1
5
7
8
1
0
1
0
5
2
0
1
5
2
0
5
2
3
3
2
1
3
2
1
1
1
1
1
6
1
0
4
4
1
0
0
0
1
1
2
1
1
2
1
3
3
1
0
3
3
1
3
1
4
2
0
1
0
2
0
3
8
1
2
2
2
2
2
1
2
14
1
3
0
10
0
0
17
2
1
1
9
0
4
3
1
2
1
1
4
6
2
6
4
2
7
0
2
1
0
1
14
0
1
0
1
10
9
1
1
2
3
4
2
4
2
0
16
4
1
5
6
0
4
1
10
1
1
4
2
6
2
9
2
1
0
4
1
6
0
7
0
0
0
17
0
0
0
2
1
0
1
4
1
1
0
1
2
2
1
2
7
2
0
9
0
2
1
0
8
0
3
5
1
1
0
3
4
3
3
3
3
1
7
9
5
9
2
2
2
1
4
0
3
5
4
7
0
2
2
6
0
0
0
3
0
1
1
2
3
1
1
4
1
2
1
0
2
5
9
7
3
1
1
7
1
1
0
3
1
3
8
1
0
1
0
3
4
1
1
3
2
7
3
0
7
2
1
6
1
3
3
3
2
1
3
2
8
2
6
3
1
1
2
9
2
6
2
2
1
5
1
3
1
0
10
2
2
2
1
8
1
2
4
1
5
4
4
1
8
0
6
2
5
1
5
2
0
0
5
2
0
1
3
1
6
2
2
3
2
1
2
2
2
0
5
1
4
0
33
5
2
3
2
2
2
3
1
0
3
4
3
2
4
1
2
2
3
3
2
4
1
4
2
3
1
2
0
14
0
18
1
3
1
4
3
0
1
9
4
3
4
2
2
2
4
3
2
9
1
3
3
8
7
1
5
1
0
5
2
3
6
0
7
2
8
1
10
1
4
0
2
2
4
0
1
0
18
6
2
0
11
1
2
2
0
4
6
2
1
1


In [ ]:
#Checking length of posts_comments_skeptics same as length of posts_skeptics:
len(posts_comments_skeptics)

In [25]:
#Instantiating dataframe to place posts, post titles, post permalinks, and post comments into:
df_posts_skeptics = pd.DataFrame()
#Adding titles_skeptics to 'title' column of df:
df_posts_skeptics['title'] = titles_skeptics
#Adding posts_skeptics to 'full_post' column of df:
df_posts_skeptics['full_post'] = posts_skeptics
#Adding permalinks_skeptics to 'permalink' column of df:
df_posts_skeptics['permalink'] = permalinks_skeptics
#Adding post_text_skeptics to 'post_text' column of df:
df_posts_skeptics['post_text'] = post_text_skeptics
#Adding posts_comments_skeptics to 'full_comments_page_each_post' column of df:
df_posts_skeptics['full_comments_page_each_post'] = posts_comments_skeptics
#Inspecting the head of the df:
df_posts_skeptics.head()

title  \
0  TIL that despite an above average sea level ri...   
1  What the alarmists won't tell you about Arctic...   
2  The Lost History of One of the World’s Strange...   
3  Pelosi’s New Climate Bill Binds America to Par...   
4  Global lightning activity and the hiatus in gl...   

                                           full_post  \
0  {'kind': 't3', 'data': {'approved_at_utc': Non...   
1  {'kind': 't3', 'data': {'approved_at_utc': Non...   
2  {'kind': 't3', 'data': {'approved_at_utc': Non...   
3  {'kind': 't3', 'data': {'approved_at_utc': Non...   
4  {'kind': 't3', 'data': {'approved_at_utc': Non...   

                                           permalink  \
0  /r/climateskeptics/comments/b8w68y/til_that_de...   
1  /r/climateskeptics/comments/b8tyid/what_the_al...   
2  /r/climateskeptics/comments/b8tsml/the_lost_hi...   
3  /r/climateskeptics/comments/b6pxcp/pelosis_new...   
4  /r/climateskeptics/comments/b8qo7d/global_ligh...   

                                           post_text  \
0                                                      
1  So the url below, from government sources, cle...   
2                                                      
3                                                      
4                                                      

                        full_comments_page_each_post  
0  [{'kind': 't1', 'data': {'subreddit_id': 't5_2...  
1  [{'kind': 't1', 'data': {'subreddit_id': 't5_2...  
2  [{'kind': 't1', 'data': {'subreddit_id': 't5_2...  
3                                                 []  
4  [{'kind': 't1', 'data': {'subreddit_id': 't5_2...

In [26]:
#Saving dataframe of r/climateskeptics/new posts, post titles, post permalinks, and post comments to a csv 
#in the data file in this repo:
df_posts_skeptics.to_csv('../data/skeptics_posts_detailed_20190403_0640.csv')

I wanted to collect more data, but I ran into time constraints.  In the future, I would like to collect more data to build a better model.  